instrumentation/servlet/servlet-common/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/servlet/common/service/ServletAndFilterAdviceHelper.java

```diff
@@ -49,23 +46,33 @@
       return;
     }
 
-    AtomicBoolean responseHandled = new AtomicBoolean(false);
-    ServletAccessor<REQUEST, RESPONSE> accessor = tracer.getServletAccessor();
-
-    // In case of async servlets wait for the actual response to be ready
-    if (accessor.isRequestAsyncStarted(request)) {
-      try {
-        accessor.addRequestAsyncListener(
-            request, new TagSettingAsyncListener<>(tracer, responseHandled, context));
-      } catch (IllegalStateException e) {
-        // org.eclipse.jetty.server.Request may throw an exception here if request became
-        // finished after check above. We just ignore that exception and move on.
-      }
+    if (mustEndOnHandlerMethodExit(tracer, request)) {
+      tracer.end(context, response);
     }
+  }
 
-    // Check again in case the request finished before adding the listener.
-    if (!accessor.isRequestAsyncStarted(request) && responseHandled.compareAndSet(false, true)) {
-      tracer.end(context, response);
+  /**
+   * Helper method to determine whether the appserver handler/servlet service/servlet filter method
+   * that started a span must also end it, even if no error was detected. Extracted as a separate
+   * method to avoid duplicating the comments on the logic behind this choice.
+   */
+  public static <REQUEST, RESPONSE> boolean mustEndOnHandlerMethodExit(
+      ServletHttpServerTracer<REQUEST, RESPONSE> tracer, REQUEST request) {
+
+    if (tracer.isAsyncListenerAttached(request)) {
+      // This request is handled asynchronously and startAsync instrumentation has already attached
+      // the listener.
+      return false;
     }
+
+    // Possible scenarios:
+    // 1) synchronously handled request
+    // 2) asynchronously handled request, but startAsync instrumentation is broken on this server
+    //    (should be caught by tests for the specific server).
+    // In case of broken startAsync instrumentation, no fallback handling for asynchronous requests
+    // should be provided as handling it in the handler/service method is  prone to race conditions
+    // (seen happening on Undertow) that may make some tests pass which should fail due to the
+    // possible race.
```

Comment:
i'm hoping we can remove this part about broken startAsync instrumentation if the above comment turns out to be fruitful

Reply:
I actually added this comment before I encountered the issue with Liberty - so its purpose is to be a hint for debugging and to make it clear to the reader why we even need a separate instrumentation in the first place.

Reply:
Rephrased the comment